### PR TITLE
contrib/database/sql: add context key to support additional tags

### LINKED ISSUE #485: contrib/database/sql: No way to set span tag dynamically
It don't have way to set span tag dynamically. 
we want to know what queries are issued for each endpoint. so want to set endpoint to the span tag.

https://github.com/DataDog/dd-trace-go/blob/a59475992ec658c954de1e89fff55b4bfef54570/contrib/database/sql/conn.go#L136-L163

### PR DESCRIPTION
**Pull request by Code-Hex**
## What

Added span tag context key for sql tracer.
It becomes possible to set the tag dynamically.

## Why

As an example, I want to know what queries are issued for each endpoint.

Fixes #485

### CODE REVIEW

**Comment by gbbr** [comment]
Hi @Code-Hex!

Thanks for your first contribution! I would like to encourage you to read our [contribution guidelines](https://github.com/DataDog/dd-trace-go/blob/v1/CONTRIBUTING.md) to learn more about contributing to this repository!

That being said, please open an issue first and clearly describe the problem you are trying to solve, along with a suggested solution (if you have one). You may use this PR as an example. We can then discuss the issue together and find the best solution. Once we do, it's fine to work on writing the actual code.

**Comment by Code-Hex** [comment]
Oh, I'm sorry. I will submit new issue 😄

**Comment by Code-Hex** [comment]
Is broken dependency on CI? 🤔

File: contrib/database/sql/conn.go [hunk 1]
```diff
@@ -133,6 +133,25 @@ type traceParams struct {
 	meta       map[string]string
 }
 
+type contextKey int
+
+const spanTagsKey contextKey = 0 // map[string]string
+
+// WithSpanTags is used if you want to set the span tag in a context.
```
Comment (gbbr) [inline contrib/database/sql/conn.go]: Can we please rephrase this a bit? Perhaps:
```go
// WithSpanTags creates a new context containing the given set of tags. They will be added
// to any query created with the returned context.
```

File: contrib/database/sql/conn.go [hunk 2]
```diff
@@ -133,6 +133,25 @@ type traceParams struct {
...
+// You can set span tag in the context like this
+//
+// ctx = WithSpanTags(ctx, map[string]string{"tag_nane":"tag_value"})
```
Comment (gbbr) [inline contrib/database/sql/conn.go]: I think the example is not needed, but if you do wish to keep it, let's make sure the following points are met:

* It is correctly formatted using `gofmt`
* There are exactly 3 spaces between the comment start (`//`) and the first character of the line of code to ensure it displays _as code_ in `godoc`
* There are no `...` and it's fully working code.

It'd be nice to either remove it, or correct it. Up to you!

File: contrib/database/sql/conn.go [hunk 3]
```diff
@@ -133,6 +133,25 @@ type traceParams struct {
...
+// db.QueryContext(ctx, stmt, ...)
+func WithSpanTags(ctx context.Context, tags map[string]string) context.Context {
+	return context.WithValue(ctx, spanTagsKey, tags)
+}
+
+// SpanTagsFromContext is used if you want to get the span tags from a context.
```
Comment (gbbr) [inline contrib/database/sql/conn.go]: We can remove this one, I think. It provides no benefit.

**Comment by gbbr** [comment]
Oh, and lastly forgot: we'll need a test for it too!

File: contrib/database/sql/conn.go (hunk 2 above)
Comment (Code-Hex) [reply contrib/database/sql/conn.go]: Removed it!!

**Comment by Code-Hex** [comment]
@gbbr finished

File: contrib/database/sql/conn_test.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+			args: args{
+				ctx:  context.Background(),
+				tags: nil,
+			},
+			wantOK: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := WithSpanTags(tt.args.ctx, tt.args.tags)
```
Comment (gbbr) [inline contrib/database/sql/conn_test.go]: Sorry if I was unclear, but what I meant was to have a full on fledged test using the mocktracer and ensuring that the created spans contain the tags in the context.

**Comment by Code-Hex** [comment]
@gbbr done

File: contrib/database/sql/conn_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// Copyright 2016-2019 Datadog, Inc.
+
+package sql
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
```
Comment (gbbr) [inline contrib/database/sql/conn_test.go]: We're using `assert` throughout, we should use the same. We don't want to introduce new dependencies.

File: contrib/database/sql/conn_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	ctxTags := map[string]string{
+		"ctx_tag1": "value1",
+		"ctx_tag2": "value2",
+		"ctx_tag3": "value3",
+	}
+	ctx := WithSpanTags(context.Background(), ctxTags)
+	query := "SELECT 1"
+	startTime := time.Date(2019, 1, 1, 1, 0, 0, 0, time.UTC)
+
+	tp := &traceParams{
```
Comment (gbbr) [inline contrib/database/sql/conn_test.go]: Not like this. We want to use the real public API, not the private methods. You can get inspired from the rest of the tests. For example here: https://github.com/DataDog/dd-trace-go/blob/a59475992ec658c954de1e89fff55b4bfef54570/contrib/internal/sqltest/sqltest.go#L83-L96

Comment (Code-Hex) [reply contrib/database/sql/conn_test.go]: I didn't look this. thank you so much :bow:

**Comment by Code-Hex** [comment]
@gbbr please review again 🙏

**Comment by gbbr** [comment]
Thanks a lot for your time @Code-Hex. I made some minor adjustments and now it looks good! We'll merge it as soon as we fix CI and release it in our next release. 🙏

**Comment by gbbr** [comment]
Seems I can't push to your remote. Maybe it's disabled? Can you please apply the following changes:
```diff
diff --git a/contrib/database/sql/conn_test.go b/contrib/database/sql/conn_test.go
index 94ac1a3..741b9db 100644
--- a/contrib/database/sql/conn_test.go
+++ b/contrib/database/sql/conn_test.go
@@ -11,11 +11,11 @@ import (
 	"log"
 	"testing"
 
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
+
 	"github.com/go-sql-driver/mysql"
 	"github.com/lib/pq"
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
-	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
 )
 
 func TestWithSpanTags(t *testing.T) {
@@ -80,14 +80,14 @@ func TestWithSpanTags(t *testing.T) {
 			ctx := WithSpanTags(context.Background(), tt.want.ctxTags)
 
 			rows, err := db.QueryContext(ctx, "SELECT 1")
-			require.NoError(t, err)
+			assert.NoError(t, err)
 			rows.Close()
 
 			spans := mt.FinishedSpans()
-			require.Len(t, spans, 1)
+			assert.Len(t, spans, 1)
 
 			span := spans[0]
-			require.Equal(t, tt.want.opName, span.OperationName())
+			assert.Equal(t, tt.want.opName, span.OperationName())
 			for k, v := range tt.want.ctxTags {
 				assert.Equal(t, v, span.Tag(k), "Value mismatch on tag %s", k)
 			}
diff --git a/contrib/database/sql/option_test.go b/contrib/database/sql/option_test.go
index 8f730a8..9c29b58 100644
--- a/contrib/database/sql/option_test.go
+++ b/contrib/database/sql/option_test.go
@@ -9,7 +9,8 @@ import (
 	"testing"
 
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/globalconfig"
-	"gotest.tools/assert"
+
+	"github.com/stretchr/testify/assert"
 )
 
 func TestAnalyticsSettings(t *testing.T) {
```

**Comment by Code-Hex** [comment]
@gbbr It fixed

File: contrib/database/sql/option_test.go [hunk 7]
```diff
@@ -8,8 +8,8 @@ package sql
 import (
 	"testing"
 
+	"github.com/stretchr/testify/assert"
```
Comment (gbbr) [inline contrib/database/sql/option_test.go]: This import needs to go last and have an empty line between dd-trace-go imports and itself.

File: contrib/database/sql/conn_test.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import (
+	"context"
+	"database/sql/driver"
+	"log"
+	"testing"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/lib/pq"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
```
Comment (gbbr) [inline contrib/database/sql/conn_test.go]: You've missed this change in my diff: this import needs to be in its own group, under `testing` and have an empty line above and below it.

We list imports in 3 groups, separated by empty lines:
1. Standard library
2. Current project (dd-trace-go)
3. External dependencies

**Comment by Code-Hex** [comment]
@gbbr :+1:

**Comment by Code-Hex** [comment]
I'm the one who should be thanking you

**Comment by gbbr** [comment]
Hmmm... This seems to have broken something in CI. Can't quite figure out what...

**Comment by Code-Hex** [comment]
@gbbr done ~
